get-deployment-metrics.py

```diff
@@ -0,0 +1,286 @@
+        "--date-filter",
+        help="Github start/end date filter",
+        dest="date_filter",
+        required=True,
+    )
+    parser.add_argument(
+        "--detailed", help="Show detailed output for each repo", action="store_true"
+    )
+    parser.add_argument(
+        "--include-manual-runs",
+        help="Include manual workflow runs in stats computations",
+        dest="include_manual_runs",
+        action="store_true",
+    )
+    parser.add_argument(
+        "--verbose", help="Turn on DEBUG logging", action="store_true", required=False
+    )
+
+    args = parser.parse_args()
+
+    log_level = logging.INFO
+
+    if args.verbose:
+        print("Verbose logging selected")
+        log_level = logging.DEBUG
+
+    # Setup some logging
+    logger = logging.getLogger()
+    logger.setLevel(log_level)
+    ch = logging.StreamHandler()
+    ch.setLevel(log_level)
+    console_formatter = logging.Formatter("%(levelname)8s: %(message)s")
+    ch.setFormatter(console_formatter)
+    logger.addHandler(ch)
+
+    # Initialize connection to Github API
+    github_handle = GitHub(token=args.github_pat, paginate=True)
+    # github_handle = GitHub(token=args.github_pat)
+
+    # Get all the repos in the org
+    # /orgs/{org}/repos
+    gh_status, repo_data = github_handle.orgs[args.org_name].repos.get()
+
+    for repo in repo_data:
+        repo_name = repo["name"]
+        repo_printed = False
+
+        logger.debug("Processing repo {}".format(repo_name))
+
+        if repo["archived"]:
+            logger.debug("repo {} is archived - skipping".format(repo_name))
+            continue
+
+        # Now for each repo, see if we have a deployment workflow matching the pattern
+        # /repos/{org}/{repo name}/actions/workflows
+        gh_status, workflow_data = github_handle.repos[args.org_name][
+            repo_name
+        ].actions.workflows.get()
+
+        for workflow in workflow_data["workflows"]:
+            workflow_success_count = 0
+            workflow_success_rate = 100
+
+            workflow_fail_count = 0
+            workflow_failure_rate = 0
+
+            workflow_avg_duration = 0
+            workflow_total_duration = 0
+
+            workflow_id = workflow["id"]
+            workflow_name = workflow["name"]
+            workflow_summary_name = workflow_name.replace(
+                " ", ""
+            )  # Dicts cannot have spaces in keys
+
+            logging.debug("Found workflow {}".format(workflow_name))
+
+            if fnmatch.fnmatch(workflow_name, args.workflow_pattern):
+                logging.debug(
+                    "workflow {} matches {}".format(
+                        workflow_name, args.workflow_pattern
+                    )
+                )
+
+                # We have a matching workflow - get the runs for it in our timeframe
+                # repos/{org}}/{repo name}/actions/workflows/{workflow id}/runs
+                gh_status, workflow_runs = (
```

Comment:
I found that you can specify a query param to target certain workflow run types: https://docs.github.com/en/rest/actions/workflow-runs?apiVersion=2022-11-28#list-workflow-runs-for-a-repository
Using multiple api calls might help simplify the below logic and counter business


so for:
merges to main (normal deploy for non-terraform): `push`
rails base image update deploys: `repository_dispatch`
user manually invoked for testing: `workflow_dispatch`

https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows

Reply:
:most_interesting:.  I had to re-write the logic to get workflows because the pagination is broken in the agithub library 🤦 .  We're really only measuring the stats on the `Deploy` workflows and the only ones we don't care about are the `repository_dispatch` ones.  Given the numbers are small (hundreds), it's easier just to get all the runs and ignore the dispatch ones (unless the flag is specified).